apps/aecore/src/aec_peers.erl

```diff
@@ -237,164 +235,146 @@ check_ping_interval_env() ->
-    end;
+    {Scheme, Host, Port} = aeu_env:local_peer(),
+    LocalPeer = #peer{scheme = Scheme, host = Host, port = Port},
+    {ok, #state{peers = gb_trees:empty(),
+                local_peer = LocalPeer}}.
+
+handle_call({set_local_peer_uri, Peer}, _From, State) ->
+    {reply, ok, State#state{local_peer = Peer}};
 handle_call(get_local_peer_uri, _From, State) ->
-    {reply, State#state.local_peer_uri, State};
-handle_call({info, PeerUri}, _From, State) ->
-    case lookup_peer(PeerUri, State) of
-        {value, _, Peer} ->
-            {reply, {ok, Peer}, State};
-        none ->
-            {reply, {error, "peer not found"}, State}
-    end;
-handle_call({is_blocked, Uri}, _From, #state{blocked = Blocked} = State) ->
-    case gb_sets:is_element(Uri, Blocked) of
-        true -> {reply, true, State};
-        false ->
-            case lookup_peer(Uri, State) of
-                {value, _, #peer{blocked = true}} ->
-                    %% shouldn't happen, but let's repair if it does
-                    Blocked1 = gb_sets:add_element(Uri, Blocked),
-                    {reply, true, State#state{blocked = Blocked1}};
-                {value, _, _} ->
-                    {reply, false, State};
-                none -> {reply, false, State}
-            end
-    end;
+    {reply, uri_of_peer(State#state.local_peer), State};
+handle_call({is_blocked, Peer}, _From, State) ->
+    {reply, is_blocked(Peer, State), State};
 handle_call(all, _From, State) ->
-    Uris = [ uri(Peer) || Peer <-  gb_trees:values(State#state.peers) ],
+    Uris = [ uri_of_peer(Peer) || Peer <- gb_trees:values(State#state.peers) ],
     {reply, Uris, State};
-handle_call({get_random, N, Exclude}, _From, State) ->
-    {reply, get_random_n(N, Exclude, State), State};
-handle_call(get_random, _From, State) ->
-    case gb_trees:is_empty(State#state.peers) of
-        true ->
-            {reply, {error, "we don't know any peers"}, State};
-        false ->
-            RandomHash = [rand:uniform(256)-1 || _ <- lists:seq(1, 16)],
-            %RandomHash = <<RandomInts:16/big-unsigned-integer-unit:8>>,  %This line generates a fairly random 16byte hex string
-            {LargestKey, _} = gb_trees:largest(State#state.peers),
-            case (binary_to_list(LargestKey) < RandomHash) of
-                true ->  %we choese an Hash after the last Hash in tree so we return the first element
-                    {_, Peer} = gb_trees:smallest(State#state.peers),
-                    {reply, {ok, Peer}, State};
-                false ->
-                    Iterator = gb_trees:iterator_from(list_to_binary(RandomHash), State#state.peers),
-                    {_, Peer, _} = gb_trees:next(Iterator),
-                    {reply, {ok, Peer}, State}
-            end
-    end.
+handle_call({get_random, N0, Exclude}, _From, #state{peers = Tree0,
+                                                     errored = Errored} = State) ->
+    %% first, remove all blocked peers
```

Comment:
"remove all **errored** peers", right? Why not the blocked too?
Ah, because blocked peers are removed from the `peers` set when blocked.

Reply:
Well spotted, the comment is wrong. The invariant that we guarantee is that Blocked peers is disjunct to peers. Not so for errored peers, because we need to remember when we latest communicated with them to be able to try again in due time. This is not stored in the URI.

Note, however, that keeping them in peers is expensive. If you have 10,000 errored peers it takes a long time each round you want to find a few none errored once.